src/librbd/crypto/FormatRequest.cc

```diff
@@ -106,8 +120,10 @@ void FormatRequest<I>::handle_flush(int r) {
 
 template <typename I>
 void FormatRequest<I>::finish(int r) {
-  if (r == 0) {
-    util::set_crypto(m_image_ctx, m_format->get_crypto());
+  ldout(m_image_ctx->cct, 20) << "r=" << r << dendl;
+
+  if (r == 0 && m_image_ctx->parent == nullptr) {
```

Comment:
The `if` body loads the encryption to the formatted image.
If this image is a child of an encrypted parent, the user will get into a weird state where encryption is only partially loaded.

Reply:
They need to be explicitly loaded by the user after the format is completed.

If we load the encryption following the format of a clone, we implicitly require the user to load the encryption before formatting (to avoid getting into the invalid encryption state).

On the other hand, my suggestion of not loading the encryption following the format of a clone can avoid the invalid state occurred above, since we also reject formatting of a cloned image with encryption loaded.

Reply:
> They need to be explicitly loaded by the user after the format is completed.
> 
> If we load the encryption following the format of a clone, we implicitly require the user to load the encryption before formatting (to avoid getting into the invalid encryption state).

Just to make sure I understand: the invalid encryption state referred to here is crypto dispatches, etc being loaded just for the top-most image and neither of the potentially encrypted ancestors, right?

If so, I agree with requiring an explicit load step but let's document it with a comment on `m_image_ctx->parent == nullptr` check here and also in `include/rbd/librbd.h` on `rbd_encryption_format`.

> 
> On the other hand, my suggestion of not loading the encryption following the format of a clone can avoid the invalid state occurred above, since we also reject formatting of a cloned image with encryption loaded.

I'm still on the fence regarding rejecting formatting of a cloned image with encryption loaded though.  Since `ShutDownCryptoRequest` would unload encryption for the entire chain and we are requiring an explicit load step for cloned images, it seems like a made-up restriction to me.

Reply:
> I'm still on the fence regarding rejecting formatting of a cloned image with encryption loaded though.
 
Scratch that -- after seeing your other comment, it makes sense to me (with a fixed up error message).